### DISCUSSION TITLE
Attacker diversity

### DISCUSSION THREAD

**Discussion by The-JMFB** [ochadenas/cpudefense#138] [Ideas]
I think one of the things that could help with the growth and development of new defensive options would be to consider new attacker types/behavior. We're defending against a virus, right? What do real computer viruses do? They adapt, they obfuscate, they masquerade as legitimate code, they subvert or disable defenses.

What if a certain type of attacker were immune to a certain type of chip?  What if a certain type of attacker could damage/destroy a certain type of defensive chip?

The introduction of this complexity would increase player interaction, increase strategic planning, and introduce the possibility of phased levels where early waves are your basic attackers and later waves begin introducing these new advanced attackers.

I would rather be spending information on defense maintenance than the constantly increasing costs of chips and their upgrades. The idea of repairing chips seems much more practical than the idea of a chip being more expensive than the last just because you already have one. I think the logic of spending resources for maintenance would be much more palatable than the idea of chips being arbitrarily more expensive. 

The idea behind the growing chip costs seems to be nothing more than an artificial balancing mechanism to prevent a player from simply buying 20 ACC chips and a SUB to beat an infinite number of attackers. That's understandable, but doesn't seem like a logical fit for the lore. If the virus can damage the chips, then a constant maintenance need will help to provide a balance and add another stratagem to the overall game. 

You could use a simple color coding to identify the type of attacker. For instance, a blue attacker could be immune to SHR chips, or maybe inflict a certain percentage of it's received damage (from the SHR chip) on the SHR chip.

I think the negative chips are similar logic, but we can remove those permanently, where the different types of attackers would induce a constant need to monitor and maintain chips. Most of the attackers could still be the normal attackers, but insert a small variety of different attacker types to later waves to ratchet up the difficulty, strategy, and player interaction in a way that seems logical and less arbitrary than using chip/upgrade costs to limit the players. 

This would obviously be a very complex change to implement, but I think it might be easier to provide balance this way than trying to constantly adjust prices as a balancing mechanism.

**Comment by Cwpute** [comment]
Im grateful someone else sees the interest in such iteration :D i've been wanting to talk about it but there are already so much discussions opened in which we go in-depth talking about the game that i thought i'd wait and see what happens.

But while it's somewhat easy to come up with interesting attacker ideas, i fear it will be hard to find interesting and elegant ways to integrate  counter-measures to their new abilities, without overwhelming the player.
Afterall, we do not have very much agency over chip placement: it seems to me the game was never built to play with and challenge your adaptability, but rather try to oppose players with an increasingly complete system of defense that you learn your way around as you play. But there is so much you can mix into to a meal until it devolves into an unrecognizable, unsavory broth…

So let's try to create interesting attackers *AND* interesting ways to deal with them :)

**Reply by Cwpute** [reply]
One of my prefered ideas was (not anymore and i'll explain why) having attackers that look like corrupted data — with constantly changing characters, you've seen this in other games— which cannot be dealt with by chips until the player taps on them. The catch here is that once the attacker's corrupted data is revealed, it will infect the next chip it reaches with its malicious code (malus effect: temporary deactivation, destruction, level decrease…), thus providing you with a chance to chose which chip you'd rather see affected by a malus you don.t know exactly about until you tap on that attacker.
The potential it provides is interesting éecause it requires you to have a comprehensive understanding of your defense to prioritize one chip over the other, and requires you act in-between chips, while most actions in the game happens on chip slots.

But i'm not sure this is a good idea anymore, because we've already got more player unteraction in the fonm of MEM chip management, and this alone might prove too much to deal with, let alone add contextual interaction on attackers.

**Comment by Cwpute** [comment]
One interesting way to create unique attackers easily would be to have a handful of them spawn with uniquely high stats… an attacker with insane speed but low power, or insane power but low speed (or uniquely long path towards the CPU)… just like a lone cryptocoin can pose a threat among a series of different other attackers.

I'd be curious to see if entires waves of characterized attackers like these, mixed in with classic ones, could throw players off and distract their perfect lines of defense… 🤔

**Comment by Cwpute** [comment]
For what it's worth, and in complete disregard of what i just said about "elegant solutions" and "interesting suggestions"… here is a list of random attacker ideas 😄 :
1. Attackers with hidden value (effectively cryptocoins, bar their current immunity for ACC and MEM)
2. Attackers that deactivate chips: the chip they reach is rendered inactive
3. Attackers that devolve chips: reduces the level of a chip by 1
4. Attackers that continuously increase in value
5. Attacker-caterpillars: several interlinked attackers that won't die until you destroy their "head", at which point they all die instantaneously

**Comment by Cwpute** [comment]
> What if a certain type of attacker were immune to a certain type of chip?

I'm much more in favor of attackers like that, which can be immune to any type of chip, but clearly defined as such, rather than our current criptocoins which try to balance ACC and MEM chips without realy adressing the core issue.
Good idea! If somehow colour-coding attackers could be accessible to colour-blind people as well!… 👀

> What if a certain type of attacker could damage/destroy a certain type of defensive chip?

Interesting, but difficult to implement i think. There is a lot to define and take into account.
Could we prevent a chip from being destroyed ? for example, would it act like TF2's Spy's sapper and set a auto-destruction timer on the chip unless the player defuse it somehow ?
Would destroying a lvl1 chip enravel in the same way as a lvl12 chip would ? would you get some info back ?
What sort of damage would be applied to the chip ? worsened cooldown ? level downgrade ? deactivation ? would ve be able to repair it back to normal ? how ? would we be able to prevent it ? how ?
It's a whole rabbit-hole of questions.

**Comment by Cwpute** [comment]
I was thinking: we already have 3 different types of attackers ! Cryptocoins, yes… and bitwise and hexadecimal attackers !

Perhaps you could have something that prevents some of your chips from interacting with either bitwise or hexadecimal attackers, or which interaction is only worth for one or the other… Special chip slots maybe, that apply their effect to the chip they are set upon ? whole copper lane sections that do the same ? chips that are hard-coded to do so, like the current ACC and MEM for the crypto ?
